worlds/tww/randomizers/Dungeons.py

```diff
@@ -0,0 +1,263 @@
+    """
+    Retrieve the item pool for all The Wind Waker dungeons in the multiworld.
+
+    :param multiworld: The MultiWorld instance.
+    :return: List of dungeon items across all The Wind Waker dungeons.
+    """
+    return [
+        item for world in multiworld.get_game_worlds("The Wind Waker") for item in get_dungeon_item_pool_player(world)
+    ]
+
+
+def get_dungeon_item_pool_player(world: "TWWWorld") -> list[Item]:
+    """
+    Retrieve the item pool for all dungeons specific to a player.
+
+    :param world: The Wind Waker game world.
+    :return: List of items in the player's dungeons.
+    """
+    return [item for dungeon in world.dungeons.values() for item in dungeon.all_items]
+
+
+def get_unfilled_dungeon_locations(multiworld: MultiWorld) -> list[Location]:
+    """
+    Retrieve all unfilled The Wind Waker dungeon locations in the multiworld.
+
+    :param multiworld: The MultiWorld instance.
+    :return: List of unfilled The Wind Waker dungeon locations.
+    """
+    return [
+        location
+        for world in multiworld.get_game_worlds("The Wind Waker")
+        for location in multiworld.get_locations(world.player)
+        if location.dungeon and not location.item
+    ]
+
+
+def modify_dungeon_location_rules(locations: list[Location], dungeon_specific: set[tuple[int, str]]) -> None:
+    """
+    Modify the rules for The Wind Waker dungeon locations based on specific player-requested constraints.
+
+    :param locations: List of dungeon locations to modify.
+    :param dungeon_specific: Set of dungeon-specific item constraints.
+    """
+    for location in locations:
+        if dungeon_specific:
+            dungeon = location.dungeon
+            orig_rule = location.item_rule
+            location.item_rule = lambda item, dungeon=dungeon, orig_rule=orig_rule: (
+                not (item.player, item.name) in dungeon_specific or item.dungeon is dungeon
+            ) and orig_rule(item)
+
+
+def fill_dungeons_restrictive(multiworld: MultiWorld) -> None:
+    """
+    Correctly fill The Wind Waker dungeons in the multiworld.
+
+    :param multiworld: The MultiWorld instance.
+    """
+    localized: set[tuple[int, str]] = set()
+    dungeon_specific: set[tuple[int, str]] = set()
+    for subworld in multiworld.get_game_worlds("The Wind Waker"):
+        player = subworld.player
+        if player not in multiworld.groups:
+            localized |= {(player, item_name) for item_name in subworld.dungeon_local_item_names}
+            dungeon_specific |= {(player, item_name) for item_name in subworld.dungeon_specific_item_names}
+
+    if localized:
+        in_dungeon_items = [item for item in get_dungeon_item_pool(multiworld) if (item.player, item.name) in localized]
+        if in_dungeon_items:
+            locations = [location for location in get_unfilled_dungeon_locations(multiworld)]
+            modify_dungeon_location_rules(locations, dungeon_specific)
```

Comment:
I think this change in rules is going to fall afoul of https://github.com/ArchipelagoMW/Archipelago/issues/4563

Either the rules would need to be reverted after the dungeon fill is complete (I think progression balancing won't currently balance locally placed items, but you might want to lock the filled locations anyway to make sure that no worlds or progression balancing in the future can then move the placed items to different locations outside the dungeon(s)), or these rules would need to be set in `set_rules()`.

Reply:
I moved the modification of dungeon rules from `stage_pre_fill` to `stage_set_rules` which hopefully resolves this issue.